# Loop excercises - Solution

## 0 - Collatz conjecture

The German mathematician Lothar Collatz came up with a fascinating (and still unproven) hypothesis. It goes like this:

If you start with any positive integer and then
- divide it by 2 if it is even
- multiply it by 3 and add 1 if it is odd

and continue to do so with the new value, you will always arrive at 1.

Write code that asks the user to input a positive integer and then performs the calculations above until the value is 1. Also, count the steps needed to reach 1 and print out the value at each step (make sure the output is an integer).

*(There is also a larger exercise around the Collatz conjecture on Moodle. Make sure to check it out once you are familiar with functions)*

In [ ]:
number = int(input())

while number != 1 and number != -1:
  print(number)
  if number % 2 != 0:
    number = number * 3 + 1
  else:
    number = number // 2

print(number)

10
10
5
16
8
4
2
1


## 1 - Perfect numbers

A perfect number is a number which is equal to the sum of its factors (excluding itself, of course). For example, 6 is a perfect number because its factors are 1, 2 and 3, which add up to 1 + 2 + 3 = 6.

Find all perfect numbers up to 10,000.

In [ ]:
def is_perfect_number(number):
    clean_divisor = []
    for i in range(1, number):
      if number % i == 0:
        clean_divisor.append(i)

    return sum(clean_divisor) == number

In [ ]:
perfect_numbers = []
for i in range(1, 10000):
  if is_perfect_number(i) == True:
    perfect_numbers.append(i)




[6, 28, 496, 8128]

You can learn more about perfect numbers [here](https://www.youtube.com/watch?v=Zrv1EDIqHkY).

## 2 - Triangular numbers

Triangular numbers are numbers that are the sum of 1+2+3+4...+n, i.e. 1, 3, 6, 10, ...

They are called triangular numbers because they make up triangles:

```
            *
       *   **
   *  **  ***
* ** *** **** ...
```

In [ ]:
"""
tri 1 = 1
tri 2 = 1 + 2
tri 3 = 1 + 2 + 3
tri 4 = 1 + 2 + 3 + 4

for number in range(1, 100):
  if number == number * (number + 1) / 2:
    print(number)


n = 1
while True:
    tri_num = n * (n + 1) // 2
    if tri_num > 100:
        break
    print(tri_num)
    n += 1
"""

tri_num = 1
n = 1
while (tri_num < 100):
  tri_num = n * (n + 1) // 2
  print(tri_num)
  n += 1

1
3
6
10
15
21
28
36
45
55
66
78
91
105


Find all triangular numbers up to 10000.

In [ ]:
triangular_numbers = [] #for the next question
tri_num = 1
counter = 2
while (tri_num < 100):
  print(tri_num)
  triangular_numbers.append(tri_num)
  tri_num += counter
  counter += 1


1
3
6
10
15
21
28
36
45
55
66
78
91


## 3 - Are perfect numbers triangular?

Quickly check with a loop whether the perfect numbers from 1 are triangular numbers (from 2).

In [ ]:
for i in range(len(perfect_numbers)):
  if perfect_numbers[1] in triangular_numbers:
    print(perfect_numbers[i])


6
28
496


## 4 - Beautiful math

Write two loops that print out the product of 12345679 and multiples of 9 up to 81 and then the product of 987654321 and again multiples of 9 up to 81.

In [ ]:
for i in range(9, 81):
  if i % 9 == 0:
    result = i * 12345679
    print(result)

111111111
222222222
333333333
444444444
555555555
666666666
777777777
888888888


In [ ]:
for i in range(9, 81):
  if i % 9 == 0:
    result = i * 987654321
    print(result)

8888888889
17777777778
26666666667
35555555556
44444444445
53333333334
62222222223
71111111112


## 5 - Beautiful math II

Write a loop that calculates (and prints out) 0 X 9 + 1, 1 X 9 + 2, 12 X 9 + 3 up to 123456789 + 9.

In [ ]:
big_value = [1,2,3,4,5,6,7,8,9]
for i in big_value:
  result23 = i * 9
  print(result23)

9
18
27
36
45
54
63
72
81


In [ ]:
big_value = ["0", "1", "2", "3", "4", "5", "6", "7", "8" ,"9"]

first = big_value[0]
second = first + big_value[1]
third = second + big_value[2]
fourth = third + big_value[3]
fifth = fourth + big_value[4]
sixth = fifth + big_value[5]
seventh = sixth + big_value[6]
eight = seventh + big_value[7]
ninth = eight + big_value[8]
tenth = ninth + big_value[9]
all_numbs = [first, second, third, fourth, fifth, sixth, seventh, eight, ninth, tenth]

big_value2 = ["1", "2", "3", "4", "5", "6", "7", "8" ,"9"]

integer_list = [int(i) for i in all_numbs]
integer_big_value = [int(i) for i in big_value2]


not_added = []
for i in integer_list:
  result2 = i * 9
  not_added.append(result2)

[x + y for x, y in zip(not_added, integer_big_value)] #found the zip function on google, very cool


[1, 11, 111, 1111, 11111, 111111, 1111111, 11111111, 111111111]

In [ ]:
big_value = ["0", "1", "2", "3", "4", "5", "6", "7", "8" ,"9"]
big_value2 = ["1", "2", "3", "4", "5", "6", "7", "8" ,"9"]

all_numbs = []
cumulative = ""
for i in big_value:
  cumulative += i #didnt realise you can concat this way
  all_numbs.append(cumulative)

integer_list = [int(i) for i in all_numbs]
integer_big_value = [int(i) for i in big_value2]


not_added = []
for i in integer_list:
  result2 = i * 9
  not_added.append(result2)

[x + y for x, y in zip(not_added, integer_big_value)]


[1, 11, 111, 1111, 11111, 111111, 1111111, 11111111, 111111111]

In [ ]:
cumulative=""
store = []
for i in range(10):
  cumulative += str(i)
  store.append(int(cumulative) * 9 + (1 + i))

print(store)

[1, 11, 111, 1111, 11111, 111111, 1111111, 11111111, 111111111, 1111111111]


In [ ]:
int("012")

12

## 6 - Beautiful math III

Write a loop that calculates (and prints out) 0 X 9 + 8, 9 X 9 + 7, 98 X 9 + 6 up to 98765432 X 9 + 0.

In [ ]:
big_value = ["1", "2", "3", "4", "5", "6", "7", "8" ,"9", "0"]
big_value2 = ["1", "2", "3", "4", "5", "6", "7", "8"]

new = reversed(big_value)
new2 = reversed(big_value2) #the curse of being lazy

all_numbs = []
cumulative = ""
for i in new:
  cumulative += i
  all_numbs.append(cumulative)

integer_list = [int(i) for i in all_numbs]
integer_big_value = [int(i) for i in new2]


not_added = []
for i in integer_list:
  result2 = i * 9
  not_added.append(result2)

[x + y for x, y in zip(not_added, integer_big_value)]

#what i did here was remove the 0 from big_value and put it at the end to get the set of values we need when reversed
#also got rid of 9 from big_value 2 because the set of numbers for the addition part starts from 8

[8, 88, 888, 8888, 88888, 888888, 8888888, 88888888]

In [14]:
cumulative = ""
store = []

for i in range(9):  # Loop runs 9 times (from 0 to 8)
    cumulative += str(9 - i)  # Constructing numbers in reversed order
    store.append(int(cumulative) * 9 + (8 - i) - 1)  # Applying the correct operation

print(store)


[88, 888, 8888, 88888, 888888, 8888888, 88888888, 888888888, 8888888888]


## 7 - Beautiful math IV

Write a loop that prints out the square of 1, 11, 111 and so on until 1111111111 (ten digits).

In [ ]:
big_one = ["1", "1", "1", "1", "1", "1", "1", "1" ,"1", "1"]
all_one = []
cumulative_one = ""
for i in big_one:
  cumulative_one += i
  all_one.append(cumulative_one)

integer_one = [int(i) for i in all_one]

for i in integer_one:
  squared = i ** 2
  print(squared)



1
121
12321
1234321
123454321
12345654321
1234567654321
123456787654321
12345678987654321
1234567900987654321


## 8 - Beautiful math V

Write a loop that calculates (and prints out) 1 X 8 + 1, 12 X 8 + 2, 123 X 8 + 3 until 123456789 X 8 + 9.

In [ ]:
big_value = ["1", "2", "3", "4", "5", "6", "7", "8" ,"9"]
big_value2 = ["1", "2", "3", "4", "5", "6", "7", "8" ,"9"]

all_numbs = []
cumulative = ""
for i in big_value:
  cumulative += i
  all_numbs.append(cumulative)

integer_list = [int(i) for i in all_numbs]
integer_big_value = [int(i) for i in big_value2]


not_added = []
for i in integer_list:
  result2 = i * 8
  not_added.append(result2)

[x + y for x, y in zip(not_added, integer_big_value)]

#all i did here was remove the '0' from big_value since the addition starts from 1
#and result2 = i * 8 and not 9 like in prev exercise


[9, 98, 987, 9876, 98765, 987654, 9876543, 98765432, 987654321]

In [17]:
cumulative = ""
store = []

for i in range(9):
    cumulative += str(i + 1)
    store.append(int(cumulative) * 8 + (i + 1))
print(store)

[9, 98, 987, 9876, 98765, 987654, 9876543, 98765432, 987654321]


## 9 - Number of digits
Write a loop that returns the number of digits on an integer.

In [ ]:
num = (220, 0)

for i in num:
  if i < 0:
    turn_positive = (i * -i)/i
    positive = int(turn_positive)
    print(len(str(positive)))
  else:
    print(len(str(i)))

#issue here is that im returni the digits of the integers, from a list of integers
#i think ill need to use a while loop for a single number but hopefully not since i dont like while loops

3
1


In [ ]:
num = -26755
while num < 0:
  num = (num * -num)/num
  num = int(num)
  print(len(str(num)))
else:
  print(len(str(num)))

#ended up doing a while loop, the for loop kind of needs a list to iterate through, can't have a single number


5
5


## 10 - Reverse number

Write a loop that reverses an integer, i.e. turns 123456 into 654321.

*Bonus: If you solved this with the use of strings, try to do it without. If you did it without the use of strings, use strings.*

In [ ]:
number = "123456"
number[::-1]


'654321'

In [ ]:
num = 1515
reversed_num = 0

while num != 0:
    digit = num % 10
    reversed_num = reversed_num * 10 + digit
    num //= 10

print(reversed_num)
#the loop takes the last digit of our num through % 10
#then reversed num * 10 and + that one last digit we removed
#since reversed num starts at 0, the last digit of num turns out to be the first value of reversed num as we want
#the num is floor divd, essentially getting rid of the last number

#next iteration, our 'num' is 1 digit less and process repeats
#since reversed num is no longer 0, the *10 actually does something
#it adds a second 0 value to the num and adds the new 'last' number to it and so on

5151


## 11 - Pattern creation

Write a loop that creates the following pattern:

`*`<br>
`* *`<br>
`* * *`<br>
`* * * *`<br>
`* * * * *`<br>
`* * * *`<br>
`* * *`<br>
`* *`<br>
`*`<br>

Make the length flexible so that you can change the size of the triangle. Code it in two loops first, then a single loop.

In [ ]:
star = ('*')
desired_size = 10
for i in range(1, desired_size + 1):
  lenght = i * star
  print(lenght)

lower_desired = desired_size - 1
for i in range(lower_desired, 0, -1):
  second_lenght = i * star
  print(second_lenght)


*
**
***
****
*****
******
*******
********
*********
**********
*********
********
*******
******
*****
****
***
**
*


In [ ]:
def chain(*iterables):
    # chain('ABC', 'DEF') --> A B C D E F
    for it in iterables:
        for element in it:
            yield element

upper_range = range(1, desired_size + 1)
lower_range = range(lower_desired, 0, -1)


concat_ranges = chain(upper_range, lower_range)
for i in concat_ranges:
  tree = i * star
  print(tree)

*
**
***
****
*****
******
*******
********
*********
**********
*********
********
*******
******
*****
****
***
**
*


In [ ]:
star = ('*')
max_rows = 13
mid_point = max_rows//2
for i in range(1, max_rows + 1):
  if i <= mid_point + 1:
    lenght = i * star
  else:
    lenght = lenght[:-1]
  print(lenght)


*
**
***
****
*****
******
*******
******
*****
****
***
**
*


## 12 - Pattern creation II

Create a loop that prints 111111111, 222222222, ..., 999999999 without resorting to any strings (!).

In [ ]:
no = (1, 2)
hello = no[0] + no[1]


print(no[1]*10**9 + no[1]*10**8 + no[1]*10**7 + no[1]*10**6 + no[1]*10**5 + no[1]*10**4 + no[1]*10**3 + no[1]*10**2 + no[1]*10**1 + no[1]*10**0)

big = []
for i in range(8, 0, -1):
  number = no[1]*10**9 + no[1]*10**i
  print(number)

2222222222
2200000000
2020000000
2002000000
2000200000
2000020000
2000002000
2000000200
2000000020


In [15]:
length = int(input())
repeating_one = (10**length - 1) // 9

for i in range(1, 10):
    print(repeating_one * i)


#flexible code, can change the lenght of the output - without the use of strings

8
11111111
22222222
33333333
44444444
55555555
66666666
77777777
88888888
99999999


## 13 - Chart positions

Find the highest chart position of singles by "Orchestral Manoeuvres in the Dark" using loops.

When you are done, find the highest charting song overall and print the artist, title and chart position.

In [ ]:
hits = {"Song 1" : {"Artist" : "Orchestral Manoeuvres in the Dark", "Title" : "Sailing on the Seven Seas", "UK Chart Position" : 3},
        "Song 2" : {"Artist" : "Orchestral Manoeuvres in the Dark", "Title" : "Maid Of Orleans", "UK Chart Position" : 4},
        "Song 3" : {"Artist" : "Orchestral Manoeuvres in the Dark", "Title" : "Enola Gay", "UK Chart Position" : 8},
        "Song 4" : {"Artist" : "Chris Rea", "Title":"The Road to Hell, Part II", "UK Chart Position" : 10},
        "Song 5" : {"Artist" : "Chris Rea", "Title" : "Auberge", "UK Chart Position" : 16},
        "Song 6" : {"Artist" : "K's Choice", "Title" : "Not an Addict", "UK Chart Position" : "-"},
        "Song 7" : {"Artist" : "The Airborne Toxic Event", "Title" : "Sometime around Midnight", "UK Chart Position" : 33},
        "Song 8" : {"Artist" : "Phosphorescent", "Title" : "Song for Zula", "UK Chart Position" : "-"},
        "Song 9" : {"Artist" : "The Magnetic Fields", "Title" : "Book of Love", "UK Chart Position" : "-"},
        "Song 10" : {"Artist" : "Stereophonics", "Title" : "Just Looking", "UK Chart Position" : 4},
        "Song 11" : {"Artist" : "Syd Matters", "Title" : "Obstacles", "UK Chart Position" : "-"},
        "Song 12" : {"Artist" : "U2", "Title" : "With or Without You", "UK Chart Position" : 4},
        "Song 12" : {"Artist" : "The Cure", "Title" : "Just Like Heaven", "UK Chart Position" : 29},
        "Song 13" : {"Artist" : "Orchestral Manoeuvres in the Dark", "Title" : "Walking on the Milky Way", "UK Chart Position" : 17},
        "Song 14" : {"Artist" : "Travis", "Title" : "Why does it always rain on me?", "UK Chart Position" : 10},
        "Song 15" : {"Artist" : "Pulp", "Title" : "Common People", "UK Chart Position" : 2},
        "Song 16" : {"Artist" : "Heather Nova", "Title" : "London Rain", "UK Chart Position" : 87},
        "Song 17" : {"Artist" : "The Turth About", "Title" : "2 a.m.", "UK Chart Position" : "-"},
        "Song 18" : {"Artist" : "Bryde", "Title" : "Silence", "UK Chart Position" : "-"},
        "Song 19" : {"Artist" : "Fleetwood Mac", "Title" : "Go Your Own Way", "UK Chart Position" : 38},
        "Song 20" : {"Artist" : "Erasure", "Title" : "A Little Respect", "UK Chart Position" : 4},
        "Song 21" : {"Artist" : "Boygenius", "Title" : "Not Strong Enough", "UK Chart Position" : "-"}
       }

In [ ]:
print(hits["Song 1"]["Artist"])

Orchestral Manoeuvres in the Dark


In [ ]:
print(hits["Song 1"]["UK Chart Position"])


3


In [ ]:
for i, obj in hits.items():
  print(i)

  for y in obj:
    print(y + ':', obj["UK Chart Position"])

In [ ]:
#for the highest Orchestral Manoeuvres in the Dark rating
positions = []
for i, obj in hits.items():
  if "UK Chart Position" in obj:
    position = obj["UK Chart Position"]
    if position != "-" and obj["Artist"] == "Orchestral Manoeuvres in the Dark":
      positions.append((obj["Artist"], int(position)))

positions.sort(key = lambda x: x[1]) #sorting by chart position

print(positions)

[('Orchestral Manoeuvres in the Dark', 3), ('Orchestral Manoeuvres in the Dark', 4), ('Orchestral Manoeuvres in the Dark', 8), ('Orchestral Manoeuvres in the Dark', 17)]


In [ ]:
#overall
positions = []

for i, obj in hits.items():
    if "UK Chart Position" in obj:
        position = obj["UK Chart Position"]
        if position != "-":
            positions.append((obj["Artist"], obj["Title"], int(position)))

# Sort by chart position
positions.sort(key=lambda x: x[2])

print(positions[0])  # Best Charting Song


('Pulp', 'Common People', 2)


## 14 - User locations

Below is data about users from a website and their country:

In [ ]:
users = {"Adam":"Germany", "Bernard": "France", "Charlie":"United Kingdom", "Daniel" : "Belgium", "Esteban" : "Spain",
         "Freddy" : "United States", "Gertrud" : "Poland", "Hamza" : "Bangladesh", "Ida" : "Slovakia",
         "Jonathan" : "Netherlands", "Kurt" : "Germany", "Leo" : "Mexico", "Mohammed" : "Egypt", "Neelja" : "Sweden",
         "Otto" : "Finland", "Paul" : "Denmark", "Queenie" : "United Kingdom", "Rudolf" : "Norway", "Siegfried" : "Germany",
         "Tito" : "Romania", "Una" : "Austria", "Valerie" : "France", "Xaiver" : "Australia", "Yve" : "France",
         "Zach" : "United States", "Angela" : "Bulgaria", "Barbara" : "Luxemburg", "Chen" : "China", "Debby" : "United States",
         "Eddy" : "Canada", "Felix" : "Germany", "Gina" : "Belgium", "Hulya" : "Turkey", "Iggy" : "United States",
         "Jules" : "Belgium", "Kim" : "Korea", "Lisa" : "Switzerland", "Mira" : "Portugal", "Nathan" : "Austria",
         "Oni" : "Japan", "Petra" : "Germany", "Qasim" : "Pakistan", "Rita" : "Hungary", "Sarah" : "Iran", "Tina" : "Ireland",
         "Umar" : "Marocco", "Xena" : "Greece", "Yana" : "Canada", "Zoe" : "Iceland"}

In [ ]:
countries = {"Germany" : "Europe", "France" : "Europe", "United Kingdom" : "Europe", "Belgium" : "Europe", "Spain" : "Europe",
             "United States" : "North America", "Poland" : "Europe", "Bangladesh" : "Asia", "Slovakia" : "Europe",
             "Netherlands" : "Europe", "Mexico" : "North America", "Egypt" : "Africa", "Sweden" : "Europe",
             "Finland" : "Europe", "Denmark" : "Europe", "Norway" : "Europe", "Romania" : "Europe", "Austria" : "Europe",
             "Australia" : "Australia", "Bulgaria" : "Europe", "Luxemburg" : "Europe", "China" : "Asia", "Korea" : "Asia",
             "Switzerland" : "Europe", "Canada" : "North America", "Turkey" : "Asia", "Portugal" : "Europe", "Japan" : "Asia",
             "Pakistan" : "Asia", "Hungary" : "Europe", "Iran" : "Asia", "Ireland" : "Europe", "Marocco" : "Africa",
             "Greece" : "Europe", "Iceland" : "Europe"}

Use loops to create a dictionary that counts the number of users per continent and print out the result (nicely). Can you write the loop in one line?

In [ ]:
"Björn" in users
#would be funny if it returned True

False

In [ ]:
counter = {}
for i, y in users.items():
  if y in countries:
    continent = countries[y]
    counter[continent] = counter.get(continent, 0) + 1

for continent, count in counter.items():
  print(f"{continent}: {count} users")




Europe: 32 users
North America: 7 users
Asia: 7 users
Africa: 2 users
Australia: 1 users


## 15 - Pattern creation III

Create the following pattern usings loops:
```
- - + - -
- + + + -
+ + + + +
- + + + -
- - + - -
```
Make it flexible so that you can easily create similar patterns of different (odd) sizes.

In [ ]:
minus = "-"
shape = ()
for i in range(0, 6):
  box = 5 * minus
  print(box)

-----
-----
-----
-----
-----
-----


In [ ]:
i = 1
desired = int(input()) #no even
shape = []
while i < desired:
  i = i + 2
  thing = "+" * i
  shape.append(thing)
if i == desired:
  i = i - 2
  new_thing = "+" * i
  shape.append(new_thing)
print(shape)

5
['+++', '+++++', '+++']


In [ ]:
def generate_shape(size):
  shape = []
  half = (size // 2) + 1  #rows for the top half including the middle

  for i in range(1, half + 1):
    pluses = (2 * i) - 1  #1, 3, 5, 7... odd sequence
    minus = (size - pluses) // 2 #split it in two so when *plus, it's even on both side
    row = ("-" * minus) + ("+" * pluses) + ("-" * minus)
    shape.append(row)

  #mirror the top half for the bottom half excluding middle
  mirror = shape[:-1][::-1]
  shape += mirror

  for line in shape:
    print(line)

#test, odd numbers only
generate_shape(5)
print()
generate_shape(7)
print()
generate_shape(11)


--+--
-+++-
+++++
-+++-
--+--

---+---
--+++--
-+++++-
+++++++
-+++++-
--+++--
---+---

-----+-----
----+++----
---+++++---
--+++++++--
-+++++++++-
+++++++++++
-+++++++++-
--+++++++--
---+++++---
----+++----
-----+-----
